haystack/nodes/retriever/base.py

```diff
@@ -268,7 +268,7 @@ def run(  # type: ignore
         scale_score: bool = None,
     ):
         if root_node == "Query":
-            if not query:
+            if query is None:
                 raise HaystackError(
                     "Must provide a 'query' parameter for retrievers in pipelines where Query is the root node."
                 )
```

Comment:
Nitpick: it would be cool to throw a strong warning (`ERROR` level) in case the query is an empty string though. I imagine very few users want to send empty queries to the retriever, as dense retriever runs are expensive.

(untested code)

```suggestion
            if query is None:
                raise HaystackError(
                    "Must provide a 'query' parameter for retrievers in pipelines where Query is the root node."
                )
            if not isinstance(query, str) or query == "":
                logger.error("The retriever received an unusual query: '%s' This query is likely to produce garbage output.", query)
```

Reply:
The problem with this is that someone that uses the FilterRetriever will get that message error when sending the query `""` when that is the right usage

Reply:
Yeah, it looks like the docstring of FilterRetriever recommends that the user passes in an empty string https://github.com/deepset-ai/haystack/blob/9582a423a2f4826a4a15529b72fea45da611b81c/haystack/nodes/retriever/sparse.py#L363

Reply:
Just... _wow_. I was not aware. Not my definition of beautiful design :facepalm: 

Alright, how about still keeping the check for the `str` type and adding a note mentioning the exception for FilterRetriever? I'll figure out that later.

Reply:
😂 it's hard to keep a nice design everywhere

So, adding a note you mean something like this? (sorry for the bad highlighting in the code suggestion, I don't know how to add that)

```suggestion
            if query is None:
                raise HaystackError(
                    "Must provide a 'query' parameter for retrievers in pipelines where Query is the root node."
                )
            if not isinstance(query, str):
                logger.error("The retriever received an unusual query: '%s' This query is likely to produce garbage output.", query)
            if query == "":
                logger.warning("The query contains an empty string. If you are using the FilterRetriever class then this is valid, otherwise this query will likely produce garbage output")
```

Reply:
I'd even remove the warning. Unfortunately that's going to be noisy for FilterRetriever users, and given that this is a problem on our side it's not fair to spam them. Without the last two lines it's perfect :+1:

Reply:
Agree, removed!